### PR TITLE
Relayer: Ika signature logic

### PR DESCRIPTION
**Pull request by sczembor**
## Description
Code containing the `nativeProcessor` and database logic for storing the nativeTx before processing by the relayer and keeping track of what has and what hasnt been processed yet. 

Closes: #31 #61 #29 

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
    
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary

### CODE REVIEW

File: ika/client.go
Comment (vuvoth) [inline ika/client.go:16]: Notes: updateLc receives a next header.

Comment (vuvoth) [reply ika/client.go:16]: This is not within this PR scope, so no worries.

File: processor/processor.go
Comment (vuvoth) [inline processor/processor.go:17]: Maybe we can alias type  message = []byte

File: processor/processor.go
Comment (vuvoth) [inline processor/processor.go:44]: Native can send multiple transactions. So we can process the array here.

File: processor/processor.go
Comment (vuvoth) [inline processor/processor.go:27]: Why is ikaClient not a pointer? I'm curious; not a suggestion.

Comment (sczembor) [reply processor/processor.go:27]: Because its an interface

File: processor/processor.go
Comment (sczembor) [reply processor/processor.go:17]: Done

File: processor/processor.go
Comment (vuvoth) [reply processor/processor.go]: Ah, yes!

File: native2ika/processor.go
Comment (sczembor) [inline native2ika/processor.go:84]: I will create an issue for it so we do not forget to update it later #55

File: dal/db.go
Comment (sczembor) [reply dal/db.go:14]: Yes, once its signed we change it to signed

File: ika/client.go
Comment (robert-zaremba) [inline ika/client.go:15]: 1. Rename this interface to `Client`
2. Rename the struct `Client` to `client`
3. the `NewClient` constructor should return interface, rather than a struct.

File: ika/client_mock.go
Comment (robert-zaremba) [inline ika/client_mock.go:11]: let's move this file to `client_mock_test.go` or to another package

File: native2ika/processor_test.go
Comment (robert-zaremba) [inline native2ika/processor_test.go:22]: we can have a helper function: `newIcaProcessor` in tests.

File: native2ika/processor_test.go
Comment (robert-zaremba) [inline native2ika/processor_test.go:64]: why not combining this test with the previous one in a single test? You can also use `t.Run(...)`.

File: native2ika/processor.go
Comment (sczembor) [reply native2ika/processor.go:38]: Done

File: ika/client.go
Comment (sczembor) [reply ika/client.go:15]: Done

File: native2ika/processor_test.go
Comment (sczembor) [reply native2ika/processor_test.go:64]: I removed the other test, it turned out irrelevant

File: ika/client_mock.go
Comment (sczembor) [reply ika/client_mock.go:12]: This is used in different packages for testing. We cannot have it in `_test.go` because it will no be visible for other packages

**Review by vuvoth — APPROVED** [review]
LGTM! Let's wait for @robert-zaremba's final review!

File: dal/db.go
Comment (sczembor) [reply dal/db.go]: Done

File: dal/db.go
Comment (sczembor) [reply dal/db.go]: Done

File: dal/db.go
Comment (sczembor) [reply dal/db.go]: Done

File: dal/db.go
Comment (robert-zaremba) [inline dal/db.go:23]:
```suggestion
    // SrId is a ika sign request ID
	SrID      uint64      `json:"tx_id"` 
```

File: dal/db.go
Comment (robert-zaremba) [inline dal/db.go:112]:
```suggestion
			sr_id INTEGER NOT NULL,
```

File: dal/db.go
Comment (robert-zaremba) [inline dal/db.go:32]:
```suggestion
	// SrId is a ika sign request ID
	SrID      uint64          `json:"tx_id"`
```

File: dal/db.go
Comment (robert-zaremba) [inline dal/db.go:198]:
```suggestion
// GetIkaTx retrieves an Ika transaction by its primary key (sr_id and ika_tx_id).
func (db *DB) GetIkaTx(signRequestID uint64, ikaTxID string) (*IkaTx, error) {
```

File: dal/db.go
Comment (robert-zaremba) [inline dal/db.go:217]:
```suggestion
// GetBitcoinTx retrieves a Bitcoin transaction by its primary key (sr_id and btc_tx_id).
func (db *DB) GetBitcoinTx(signRequestID uint64, btcTxID []byte) (*BitcoinTx, error) {
```

File: dal/db.go
Comment (robert-zaremba) [inline dal/db.go:260]:
```suggestion
		return nil, 0, fmt.Errorf("failed to query IkaSignRequest & IkaTxStatus: %w", err)
```

**Review by robert-zaremba — APPROVED** [review]
looks good, added few suggestions. pre-approving

File: dal/db.go
Comment (sczembor) [reply dal/db.go:95]: same as above

File: native2ika/processor.go
Comment (sczembor) [reply native2ika/processor.go:37]: Done

File: nbtc/relayer_test.go
Comment (sczembor) [reply nbtc/relayer_test.go:32]: Done

File: native2ika/processor_test.go
Comment (sczembor) [reply native2ika/processor_test.go:43]: #70

File: dal/db.go
Comment (robert-zaremba) [reply dal/db.go:16]: Well, both are signatures - so should be the same type. But let's keep according to the API, and we can adapt later

File: dal/db.go
Comment (robert-zaremba) [reply dal/db.go:278]: Ok, then let's add this comment - it's valuable

File: dal/db.go
Comment (sczembor) [inline dal/db.go:269]:
```suggestion
// and are due to be broadcasted to bitcoin.
//
// This function checks for the following conditions:
// - The IkaSignRequest must have a final signature (final_sig IS NOT NULL).
// - There must be no corresponding entry in the bitcoin_txs table, OR
// - There must be only one corresponding entry in the bitcoin_txs table with a status of "Pending".
//
// The reason for checking these conditions is that we cannot have a Bitcoin transaction hash (btc_tx_id)
// before the first broadcast attempt.
```

File: dal/db.go
Comment (sczembor) [reply dal/db.go:286]: Done

**Pull request merged**
